documentation/rpio.1
.TH "RPIO" "1" "March 01, 2013" "0.8.4" "RPIO"
.SH NAME
rpio \- RPIO Documentation
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.\" Man page generated from reStructuredText.
.
.sp
RPIO is an advanced GPIO module for the Raspberry Pi.
.INDENT 0.0
.IP \(bu 2
GPIO Input and Output (drop\-in replacement for \fI\%RPi.GPIO\fP)
.IP \(bu 2
GPIO Interrupts (callbacks when events occur on input gpios)
.IP \(bu 2
TCP Socket Interrupts (callbacks when tcp socket clients send data)
.IP \(bu 2
Well documented, tested, fast source code
.IP \(bu 2
Minimal CPU and memory profile
.UNINDENT
.sp
RPIO consists of two main components:
.INDENT 0.0
.IP \(bu 2
\fI\%RPIO.py\fP, which you can import in Python 2 or 3 with \fBimport RPIO\fP
.IP \(bu 2
\fI\%rpio\fP, command\-line tools for inspecting and manipulating GPIOs system\-wide
.UNINDENT
.sp
News
.INDENT 0.0
.IP \(bu 2
v0.8.4: \fBrpio\-curses\fP
.IP \(bu 2
v0.8.2: Socket server callbacks with \fI\%RPIO.add_tcp_callback(port, callback, threaded_callback=False)\fP
.UNINDENT
.sp
The easiest way to install/update RPIO on a Raspberry Pi is with either \fBeasy_install\fP or \fBpip\fP:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
$ sudo apt\-get install python\-setuptools
$ sudo easy_install \-U RPIO
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
You can also get RPIO from Github repository, which is usually a step ahead of pypi:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
$ git clone https://github.com/metachris/RPIO.git
$ cd RPIO
$ sudo python setup.py install
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Or from Github but without Git:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
$ curl \-L https://github.com/metachris/RPIO/archive/master.tar.gz | tar \-xz
$ cd RPIO\-master
$ sudo python setup.py install
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
After the installation you can use \fBimport RPIO\fP as well as the command\-line tool \fBrpio\fP.
.sp
\fBrpio\fP includes two command\-line tools which allow you to inspect and manipulate GPIO\(aqs system wide,
including those used by other processes:
.INDENT 0.0
.IP \(bu 2
\fBrpio\fP \- command line program
.IP \(bu 2
\fBrpio\-curses\fP \- rpio with a graphical user interface for the terminal
.UNINDENT
.SH RPIO-CURSES
.sp
\fBrpio\-curses\fP is a graphical user interface for the terminal to inspect and manipulate GPIOs. Its part of RPIO since
version 0.8.4. With \fBrpio\-curses\fP you can see all GPIOs on the board with function and state, and change every
possible parameter. The list is updated every second, and instantly if you change something. Here are a few screenshots
of it in action:
.INDENT 0.0
.IP \(bu 2
\fI\%http://imgur.com/rOl9VvG\fP
.IP \(bu 2
\fI\%http://imgur.com/iB1LkW6\fP
.IP \(bu 2
\fI\%http://imgur.com/lhRvjIl\fP
.UNINDENT
.SH RPIO
.sp
\fBrpio\fP needs to run with superuser privileges (root), else it will restart using \fBsudo\fP. The BCM GPIO numbering
scheme is used by default. \fBrpio \-\-inspect\-all\fP (or \fB\-I\fP) is the most popular command; it shows you all gpios
on the board, with function and state. Here is an overview of all the functions:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
Show the help page:

    $ rpio \-h

Inspect the function and state of gpios (with \-i/\-\-inspect):

    $ rpio \-i 7
    $ rpio \-i 7,8,9
    $ rpio \-i 1\-9

    # Example output for \(garpio \-i 1\-9\(ga (non\-existing are ommitted):
    GPIO 2: ALT0   (1)
    GPIO 3: ALT0   (1)
    GPIO 4: INPUT  (0)
    GPIO 7: OUTPUT (0)
    GPIO 8: INPUT  (1)
    GPIO 9: INPUT  (0)

Inspect all GPIO\(aqs on this board (with \-I/\-\-inspect\-all):

    $ rpio \-I

Set GPIO 7 output to \(ga1\(ga (or \(ga0\(ga) (with \-s/\-\-set):

    $ rpio \-s 7:1

    You can only write to pins that have been set up as OUTPUT. You can
    set this yourself with \(ga\-\-setoutput <gpio\-id>\(ga.

Wait for interrupt events on GPIOs (with \-w/\-\-wait_for_interrupts). You
can specify an edge (eg. \(ga:rising\(ga; default=\(aqboth\(aq) as well as \(ga:pullup\(ga,
\(ga:pulldown\(ga or \(gapulloff\(ga.

    $ rpio \-w 7
    $ rpio \-w 7:rising
    $ rpio \-w 7:falling:pullup

    $ rpio \-w 7:rising:pullup,17,18
    $ rpio \-w 1\-9

Setup a pin as INPUT (optionally with software resistor):

    $ rpio \-\-setinput 7
    $ rpio \-\-setinput 7:pullup
    $ rpio \-\-setinput 7:pulldown

Setup a pin as OUTPUT (optionally with an initial value (0 or 1)):

    $ rpio \-\-setoutput 8
    $ rpio \-\-setoutput 8:1

Show Raspberry Pi system info:

    $ rpio \-\-sysinfo

    # Example output:
    000e: Model B, Revision 2.0, RAM: 256 MB, Maker: Sony
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
You can update the \fBRPIO\fP package to the latest version:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
$ rpio \-\-update\-rpio
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Install (and update) the \fBrpio\fP manpage:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
$ rpio \-\-update\-man
$ man rpio
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
RPIO.py extends \fI\%RPi.GPIO\fP in
various ways, and uses the BCM GPIO numbering scheme by default.
.INDENT 0.0
.IP \(bu 2
\fI\%GPIO Interrupts\fP
.IP \(bu 2
\fI\%TCP Socket Interrupts\fP
.IP \(bu 2
\fI\%GPIO Input & Output\fP
.IP \(bu 2
\fI\%more\fP
.UNINDENT
.SH GPIO INTERRUPTS
.sp
Interrupts are used to receive notifications from the kernel when GPIO state
changes occur. Advantages include minimized cpu consumption, very fast
notification times, and the ability to trigger on specific edge transitions
(\fBrising\fP, \fBfalling\fP or \fBboth\fP). You can also set a software pull\-up
or pull\-down resistor.
.INDENT 0.0
.TP
.B RPIO.add_interrupt_callback(gpio_id, callback, edge=\(aqboth\(aq, pull_up_down=RPIO.PUD_OFF, threaded_callback=False)
Adds a callback to receive notifications when a GPIO changes it\(aqs value. Possible \fBpull_up_down\fP values are
\fBRPIO.PUD_UP\fP, \fBRPIO.PUD_DOWN\fP and \fBRPIO.PUD_OFF\fP (default). Possible edges are \fBrising\fP,
\fBfalling\fP and \fBboth\fP (default). Note that \fBrising\fP and \fBfalling\fP edges may receive values
not corresponding to the edge, so be sure to double check.
.UNINDENT
.SH TCP SOCKET INTERRUPTS
.sp
Its easy to open ports for incoming TCP connections with just this one method:
.INDENT 0.0
.TP
.B RPIO.add_tcp_callback(port, callback, threaded_callback=False)
Adds a socket server callback, which will be started when a connected socket client sends something. This is implemented
by RPIO creating a TCP server socket at the specified port. Incoming connections will be accepted when \fBRPIO.wait_for_interrupts()\fP runs.
The callback must accept exactly two parameters: socket and message (eg. \fBdef callback(socket, msg)\fP). The callback can use the socket parameter to send values back to the client (eg. \fBsocket.send("hi there\en")\fP).
.sp
You can test the TCP socket interrupts with \fB$ telnet <your\-ip> <your\-port>\fP (eg. \fB$ telnet localhost 8080\fP). An empty string
tells the server to close the client connection (for instance if you just press enter in telnet, you\(aqll get disconnected).
.UNINDENT
.SH EXAMPLE
.sp
The following example shows how to react to events on three gpios, and one socket
server on port 8080:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
import RPIO

def gpio_callback(gpio_id, val):
    print("gpio %s: %s" % (gpio_id, val))

def socket_callback(socket, val):
    print("socket %s: \(aq%s\(aq" % (socket.fileno(), val))
    socket.send("echo: %s\en" % val)

# Three GPIO interrupt callbacks
RPIO.add_interrupt_callback(7, gpio_callback)
RPIO.add_interrupt_callback(8, gpio_callback, edge=\(aqrising\(aq)
RPIO.add_interrupt_callback(9, gpio_callback, pull_up_down=RPIO.PUD_UP)

# One TCP socket server callback on port 8080
RPIO.add_tcp_callback(8080, socket_callback)

# Start the blocking epoll loop, and catch Ctrl+C KeyboardInterrupt
try:
    RPIO.wait_for_interrupts()
except KeyboardInterrupt:
    RPIO.cleanup_interrupts()
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
If you want to receive a callback inside a Thread (to not block RPIO from returning to wait
for interrupts), set \fBthreaded_callback\fP to \fBTrue\fP when adding it:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
# for GPIO interrupts
RPIO.add_interrupt_callback(7, do_something, threaded_callback=True)

# for socket interrupts
RPIO.add_tcp_callback(8080, socket_callback, threaded_callback=True)
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
To stop the \fBwait_for_interrupts()\fP loop you can call \fBRPIO.stop_waiting_for_interrupts()\fP.
After using \fBRPIO.wait_for_interrupts()\fP you should call \fBRPIO.cleanup_interrupts()\fP before your
program quits, to shut everything down nicely.
.SH LOG OUTPUT
.sp
To enable RPIO log output, import \fBlogging\fP and set the loglevel to \fBDEBUG\fP before importing RPIO:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
import logging
log_format = \(aq%(levelname)s | %(asctime)\-15s | %(message)s\(aq
logging.basicConfig(format=log_format, level=logging.DEBUG)
import RPIO
.ft P
.fi
.UNINDENT
.UNINDENT
.SH GPIO INPUT & OUTPUT
.sp
RPIO extends \fI\%RPi.GPIO\fP;
all the input and output handling works just the same:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
import RPIO

# set up input channel without pull\-up
RPIO.setup(7, RPIO.IN)

# set up input channel with pull\-up control. Can be
# PUD_UP, PUD_DOWN or PUD_OFF (default)
RPIO.setup(7, RPIO.IN, pull_up_down=RPIO.PUD_UP)

# read input from gpio 7
input_value = RPIO.input(7)

# set up GPIO output channel
RPIO.setup(8, RPIO.OUT)

# set gpio 8 to high
RPIO.output(8, True)

# set up output channel with an initial state
RPIO.setup(8, RPIO.OUT, initial=RPIO.LOW)

# change to BOARD numbering schema
RPIO.setmode(RPIO.BOARD)

# set software pullup on channel 17
RPIO.set_pullupdn(17, RPIO.PUD_UP)  # new in RPIO

# get the function of channel 8
RPIO.gpio_function(8)

# reset every channel that has been set up by this program,
# and unexport interrupt gpio interfaces
RPIO.cleanup()
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
You can use RPIO as a drop\-in replacement for RPi.GPIO in your existing code like this:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
import RPIO as GPIO  # (if you\(aqve previously used \(gaimport RPi.GPIO as GPIO\(ga)
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
To find out more about the methods and constants in RPIO you can run \fB$ sudo pydoc RPIO\fP, or
use the help method inside Python:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
import RPIO
help(RPIO)
.ft P
.fi
.UNINDENT
.UNINDENT
.SH ADDITIONS TO RPI.GPIO
.sp
Additional Constants
.INDENT 0.0
.IP \(bu 2
\fBRPIO.RPI_REVISION\fP \- the current board\(aqs revision (either \fB1\fP or \fB2\fP)
.IP \(bu 2
\fBRPIO.RPI_REVISION_HEX\fP \- the cpu hex revision code (\fB0002\fP .. \fB000f\fP)
.UNINDENT
.sp
Additional Methods
.INDENT 0.0
.IP \(bu 2
\fBRPIO.gpio_function(gpio_id)\fP \- returns the current setup of a gpio (\fBIN, OUT, ALT0\fP)
.IP \(bu 2
\fBRPIO.set_pullupdn(gpio_id, pud)\fP \- set a pullup or \-down resistor on a GPIO
.IP \(bu 2
\fBRPIO.forceinput(gpio_id)\fP \- reads the value of any gpio without needing to call setup() first
.IP \(bu 2
\fBRPIO.forceoutput(gpio_id, value)\fP \- writes a value to any gpio without needing to call setup() first
(\fBwarning\fP: this can potentially harm your Raspberry)
.IP \(bu 2
\fBRPIO.sysinfo()\fP \- returns \fB(hex_rev, model, revision, mb\-ram and maker)\fP of this Raspberry
.IP \(bu 2
\fBRPIO.version()\fP \- returns \fB(version_rpio, version_cgpio)\fP
.UNINDENT
.sp
Interrupt Handling
.INDENT 0.0
.IP \(bu 2
\fBRPIO.add_interrupt_callback(gpio_id, callback, edge=\(aqboth\(aq, pull_up_down=RPIO.PUD_OFF, threaded_callback=False)\fP
.IP \(bu 2
\fBRPIO.add_tcp_callback(port, callback, threaded_callback=False)\fP
.IP \(bu 2
\fBRPIO.del_interrupt_callback(gpio_id)\fP
.IP \(bu 2
\fBRPIO.wait_for_interrupts(epoll_timeout=1)\fP
.IP \(bu 2
\fBRPIO.stop_waiting_for_interrupts()\fP
.IP \(bu 2
implemented with \fBepoll\fP
.UNINDENT
.sp
Please send feedback and ideas to \fI\%chris@linuxuser.at\fP, and \fI\%open an issue at Github\fP if
you\(aqve encountered a bug.
.sp
\fBHow does RPIO work?\fP
.INDENT 0.0
.INDENT 3.5
RPIO extends RPi.GPIO, a GPIO controller written in C which uses a low\-level memory interface. Interrupts are
implemented  with \fBepoll\fP via \fB/sys/class/gpio/\fP. For more detailled information take a look at the \fI\%source\fP, it\(aqs well documented and easy to build.
.UNINDENT
.UNINDENT
.sp
\fBShould I update RPIO often?\fP
.INDENT 0.0
.INDENT 3.5
Yes, because RPIO is getting better by the day. You can use \fB$ rpio \-\-update\-rpio\fP or see \fI\%Installation\fP for more information about methods to update.
.UNINDENT
.UNINDENT
.sp
\fBI\(aqve encountered a bug, what next?\fP
.INDENT 0.0
.INDENT 3.5
.INDENT 0.0
.IP \(bu 2
Make sure you are using the latest version of RPIO (see \fI\%Installation\fP)
.IP \(bu 2
Open an issue at Github
.INDENT 2.0
.IP \(bu 2
Go to \fI\%https://github.com/metachris/RPIO/issues/new\fP
.IP \(bu 2
Describe the problem and steps to replicate
.IP \(bu 2
Add the output of \fB$ rpio \-\-version\fP and \fB$ rpio \-\-sysinfo\fP
.UNINDENT
.UNINDENT
.UNINDENT
.UNINDENT
.sp
\fBpip is throwing an error during the build:\fP \fBsource/c_gpio/py_gpio.c:9:20: fatal error: Python.h: No such file or directory\fP
.INDENT 0.0
.INDENT 3.5
You need to install the \fBpython\-dev\fP package (eg. \fB$ sudo apt\-get install python\-dev\fP), or use \fBeasy_install\fP (see \fI\%Installation\fP).
.UNINDENT
.UNINDENT
.INDENT 0.0
.IP \(bu 2
\fI\%https://github.com/metachris/RPIO\fP
.IP \(bu 2
\fI\%http://pypi.python.org/pypi/RPIO\fP
.IP \(bu 2
\fI\%http://pypi.python.org/pypi/RPi.GPIO\fP
.IP \(bu 2
\fI\%http://www.raspberrypi.org/wp-content/uploads/2012/02/BCM2835-ARM-Peripherals.pdf\fP
.IP \(bu 2
\fI\%http://www.kernel.org/doc/Documentation/gpio.txt\fP
.UNINDENT
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
RPIO is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

RPIO is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
.ft P
.fi
.UNINDENT
.UNINDENT
.INDENT 0.0
.IP \(bu 2
v0.8.4
.INDENT 2.0
.IP \(bu 2
\fBrpio\-curses\fP
.IP \(bu 2
Bugfix in RPIO: tcp callbacks (first parameter \fBsocket\fP works now)
.IP \(bu 2
Renamed \fBRPIO.rpi_sysinfo()\fP to \fBRPIO.sysinfo\fP
.UNINDENT
.IP \(bu 2
v0.8.3: pypi release update with updated documentation and bits of refactoring
.IP \(bu 2
v0.8.2
.INDENT 2.0
.IP \(bu 2
Added TCP socket callbacks
.IP \(bu 2
\fBRPIO\fP does not auto\-clean interfaces on exceptions anymore, but will auto\-clean them
as needed. This means you should now call \fBRPIO.cleanup_interrupts()\fP to properly close
the sockets and unexport the interfaces.
.IP \(bu 2
Renamed \fBRPIO.rpi_sysinfo()\fP to \fBRPIO.sysinfo()\fP
.UNINDENT
.IP \(bu 2
v0.8.0
.INDENT 2.0
.IP \(bu 2
Improved auto\-cleaning of interrupt interfaces
.IP \(bu 2
BOARD numbering scheme support for interrupts
.IP \(bu 2
Support for software pullup and \-down resistor with interrupts
.IP \(bu 2
New method \fBRPIO.set_pullupdn(..)\fP
.IP \(bu 2
\fBrpio\fP now supports P5 header gpios (28, 29, 30, 31) (only in BCM mode)
.IP \(bu 2
Tests added in \fBsource/run_tests.py\fP and \fBfabfile.py\fP
.IP \(bu 2
Major refactoring of C GPIO code
.IP \(bu 2
Various minor updates and fixes
.UNINDENT
.IP \(bu 2
v0.7.1
.INDENT 2.0
.IP \(bu 2
Refactoring and cleanup of c_gpio
.IP \(bu 2
Added new constants and methods (see documentation above)
.IP \(bu 2
Bugfixes
.INDENT 2.0
.IP \(bu 2
\fBwait_for_interrupts()\fP now auto\-cleans interfaces when an exception occurs. Before you needed to call \fBRPIO.cleanup()\fP manually.
.UNINDENT
.UNINDENT
.IP \(bu 2
v0.6.4
.INDENT 2.0
.IP \(bu 2
Python 3 bugfix in \fIrpio\fP
.IP \(bu 2
Various minor updates
.UNINDENT
.UNINDENT
.SH AUTHOR
Chris Hager <chris@linuxuser.at>
.SH COPYRIGHT
2013, Chris Hager <chris@linuxuser.at>
.\" Generated by docutils manpage writer.
.


source/RPIO.py
"""
RPIO extends RPi.GPIO with interrupt handling. Importing this also sets
the default mode to GPIO.BCM (the same numbering system the kernel uses,
as opposed to the pin ids (GPIO.BOARD)).

You can use RPIO the same way as RPi.GPIO (eg. RPIO.setmode(...),
RPIO.input(...)), as well as access the new interrupt handling methods.
The following example shows how to react on events on 3 pins by using
interrupts, each with different edge detections:

    # Setup logging
    import logging
    log_format = '%(levelname)s | %(asctime)-15s | %(message)s'
    logging.basicConfig(format=log_format, level=logging.DEBUG)

    # Get started
    import RPIO

    def do_something(gpio_id, value):
        logging.info("New value for GPIO %s: %s" % (gpio_id, value))

    RPIO.add_interrupt_callback(7, do_something)
    RPIO.add_interrupt_callback(8, do_something, edge='rising')
    RPIO.add_interrupt_callback(9, do_something, pull_up_down=RPIO.PUD_UP)
    RPIO.wait_for_interrupts()

Default edge is `both` and default pull_up_down is `RPIO.PUD_OFF`.

If you want to receive a callback inside a Thread (which won't block anything
else on the system), set `threaded_callback` to True when adding an interrupt-
callback. Here is an example:

    RPIO.add_interrupt_callback(7, do_something, threaded_callback=True)

Make sure to double-check the value returned from the interrupt, since it
is not necessarily corresponding to the edge (eg. 0 may come in as value,
even if edge="rising"). To remove all callbacks from a certain gpio pin, use
`RPIO.del_interrupt_callback(gpio_id)`. To stop the `wait_for_interrupts()`
loop you can call `RPIO.stop_waiting_for_interrupts()`.

Besides the interrupt handling, you can use RPIO just as RPi.GPIO:

    import RPIO

    # set up input channel without pull-up
    RPIO.setup(7, RPIO.IN)

    # set up input channel with pull-up control. Can be
    # PUD_UP, PUD_DOWN or PUD_OFF (default)
    RPIO.setup(7, RPIO.IN, pull_up_down=RPIO.PUD_UP)

    # read input from gpio 7
    input_value = RPIO.input(7)

    # set up GPIO output channel
    RPIO.setup(8, RPIO.OUT)

    # set gpio 8 to high
    RPIO.output(8, True)

    # set up output channel with an initial state
    RPIO.setup(8, RPIO.OUT, initial=RPIO.LOW)

    # change to BOARD numbering schema
    RPIO.setmode(RPIO.BOARD)

    # set software pullup on channel 17
    RPIO.set_pullupdn(17, RPIO.PUD_UP)

    # reset every channel that has been set up by this program,
    # and unexport interrupt gpio interfaces
    RPIO.cleanup()

You can use RPIO as a drop-in replacement for RPi.GPIO in your existing code:

    import RPIO as GPIO  # (if you've used `import RPi.GPIO as GPIO`)

Author: Chris Hager <chris@linuxuser.at>
URL: https://github.com/metachris/RPIO
License: GPLv3
"""
import select
import os.path

from logging import debug, info, warn, error
from threading import Thread
from functools import partial

from GPIO import *
from GPIO import cleanup as _cleanup_orig
from GPIO import setmode as _setmode

VERSION = "0.8.0"

# BCM numbering mode by default
setmode(BCM)

# Interrupt callback maps
_map_fileno_to_file = {}
_map_fileno_to_gpioid = {}
_map_gpioid_to_fileno = {}
_map_gpioid_to_callbacks = {}

# Keep track of created kernel interfaces for later cleanup
_gpio_kernel_interfaces_created = []

# Whether to continue the epoll loop or quit at next chance. You
# can manually set this to False to stop `wait_for_interrupts()`.
_is_waiting_for_interrupts = False

# Internals
_epoll = select.epoll()
_SYS_GPIO_ROOT = '/sys/class/gpio/'
GPIO_FUNCTIONS = {0: "OUTPUT", 1: "INPUT", 4: "ALT0", 7: "-"}
MODEL_DATA = {
    '2': ('B', '1.0', 256, '?'),
    '3': ('B', '1.0', 256, '?'),
    '4': ('B', '2.0', 256, 'Sony'),
    '5': ('B', '2.0', 256, 'Qisda'),
    '6': ('B', '2.0', 256, 'Egoman'),
    '7': ('A', '2.0', 256, 'Egoman'),
    '8': ('A', '2.0', 256, 'Sony'),
    '9': ('A', '2.0', 256, 'Qisda'),
    'd': ('B', '2.0', 512, 'Egoman'),
    'e': ('B', '2.0', 512, 'Sony'),
    'f': ('B', '2.0', 512, 'Qisda')
}
_PULL_UPDN = ("PUD_OFF", "PUD_DOWN", "PUD_UP")

# List of valid bcm gpio ids for raspberry rev1 and rev2. Used for inspect-all.
GPIO_LIST_R1 = (0, 1, 4, 7, 8, 9, 10, 11, 14, 15, 17, 18, 21, 22, 23, 24, 25)
GPIO_LIST_R2 = (2, 3, 4, 7, 8, 9, 10, 11, 14, 15, 17, 18, 22, 23, 24, 25, \
        27, 28, 29, 30, 31)

# List of board pins with extra information which board header they belong to.
# Revision 2 boards have extra gpios on the P5 header (gpio 27-31)). Shifting
# the header info left by 8 bits leaves 255 possible channels per header. This
# list of board pins is currently only used for testing purposes.
HEADER_P1 = 0 << 8
HEADER_P5 = 5 << 8
PIN_LIST = (3, 5, 7, 8, 10, 11, 12, 13, 15, 16, 18, 19, 21, 22, 23, 24, 26, \
        3 | HEADER_P5, 4 | HEADER_P5, 5 | HEADER_P5, 6 | HEADER_P5)


def _threaded_callback(callback, *args):
    """ Internal wrapper to start a callback in threaded mode """
    Thread(target=callback, args=args).start()


def rpi_sysinfo():
    """ Returns (model, revision, mb-ram and maker) for this raspberry """
    return MODEL_DATA[RPI_REVISION_HEX.lstrip("0")]


def add_interrupt_callback(gpio_id, callback, edge='both',
        pull_up_down=PUD_OFF, threaded_callback=False):
    """
    Add a callback to be executed when the value on 'gpio_id' changes to
    the edge specified via the 'edge' parameter (default='both').

    `pull_up_down` can be set to `RPIO.PUD_UP`, `RPIO.PUD_DOWN`, and
    `RPIO.PUD_OFF`.

    If `threaded_callback` is True, the callback will be started
    inside a Thread.
    """
    gpio_id = channel_to_gpio(gpio_id)
    debug("Adding callback for GPIO %s" % gpio_id)
    if not edge in ["falling", "rising", "both", "none"]:
        raise AttributeError("'%s' is not a valid edge." % edge)

    if not pull_up_down in [PUD_UP, PUD_DOWN, PUD_OFF]:
        raise AttributeError("'%s' is not a valid pull_up_down value." % edge)

    # Make sure the gpio_id is valid
    if not gpio_id in (GPIO_LIST_R1 if RPI_REVISION == 1 else GPIO_LIST_R2):
        raise AttributeError("GPIO %s is not a valid gpio-id for this board." \
                % gpio_id)

    # Require INPUT pin setup; and set the correct PULL_UPDN
    if gpio_function(int(gpio_id)) == IN:
        set_pullupdn(gpio_id, pull_up_down)
    else:
        debug("- changing gpio function from %s to INPUT" % \
                (GPIO_FUNCTIONS[gpio_function(int(gpio_id))]))
        setup(gpio_id, IN, pull_up_down)

    # Prepare the callback (wrap in Thread if needed)
    cb = callback if not threaded_callback else \
            partial(_threaded_callback, callback)

    # Prepare the /sys/class path of this gpio
    path_gpio = "%sgpio%s/" % (_SYS_GPIO_ROOT, gpio_id)

    # If initial callback for this GPIO then set everything up. Else make sure
    # the edge detection is the same and add this to the callback list.
    if gpio_id in _map_gpioid_to_callbacks:
        with open(path_gpio + "edge", "r") as f:
            e = f.read().strip()
            if e != edge:
                raise AttributeError(("Cannot add callback for gpio %s:"
                        " edge detection '%s' not compatible with existing"
                        " edge detection '%s'.") % (gpio_id, edge, e))

        # Check whether edge is the same, else throw Exception
        debug("- kernel interface already configured for GPIO %s" % gpio_id)
        _map_gpioid_to_callbacks[gpio_id].append(cb)

    else:
        # If kernel interface already exists, unexport first for clean setup
        if os.path.exists(path_gpio):
            with open(_SYS_GPIO_ROOT + "unexport", "w") as f:
                f.write("%s" % gpio_id)

        # Export kernel interface /sys/class/gpio/gpioN
        with open(_SYS_GPIO_ROOT + "export", "w") as f:
            f.write("%s" % gpio_id)
        global _gpio_kernel_interfaces_created
        _gpio_kernel_interfaces_created.append(gpio_id)
        debug("- kernel interface created for GPIO %s" % gpio_id)

        # Configure gpio as input
        with open(path_gpio + "direction", "w") as f:
            f.write("in")

        # Configure gpio edge detection
        with open(path_gpio + "edge", "w") as f:
            f.write(edge)

        debug(("- kernel interface configured for GPIO %s "
                "(edge='%s', pullupdn=%s)") % (gpio_id, edge, \
                _PULL_UPDN[pull_up_down]))

        # Open the gpio value stream and read the initial value
        f = open(path_gpio + "value", 'r')
        val_initial = f.read().strip()
        debug("- inital gpio value: %s" % val_initial)
        f.seek(0)

        # Add callback info to the mapping dictionaries
        _map_fileno_to_file[f.fileno()] = f
        _map_fileno_to_gpioid[f.fileno()] = gpio_id
        _map_gpioid_to_fileno[gpio_id] = f.fileno()
        _map_gpioid_to_callbacks[gpio_id] = [cb]

        # Add to epoll
        _epoll.register(f.fileno(), select.EPOLLPRI | select.EPOLLERR)


def del_interrupt_callback(gpio_id):
    """ Delete all interrupt callbacks from a certain gpio """
    debug("- removing interrupts on gpio %s" % gpio_id)
    gpio_id = channel_to_gpio(gpio_id)
    fileno = _map_gpioid_to_fileno[gpio_id]

    # 1. Remove from epoll
    _epoll.unregister(fileno)

    # 2. Cache the file
    f = _map_fileno_to_file[fileno]

    # 3. Remove from maps
    del _map_fileno_to_file[fileno]
    del _map_fileno_to_gpioid[fileno]
    del _map_gpioid_to_fileno[gpio_id]
    del _map_gpioid_to_callbacks[gpio_id]

    # 4. Close file last. In case of IOError everything else has been shutdown
    f.close()


def _handle_interrupt(fileno, val):
    """ Internally distributes interrupts to all attached callbacks """
    gpio_id = _map_fileno_to_gpioid[fileno]
    if gpio_id in _map_gpioid_to_callbacks:
        for cb in _map_gpioid_to_callbacks[gpio_id]:
            # Start the callback!
            cb(gpio_id, val)


def wait_for_interrupts(epoll_timeout=1):
    """
    Blocking loop to listen for GPIO interrupts and distribute them to
    associated callbacks. epoll_timeout is an easy way to shutdown the
    blocking function. Per default the timeout is set to 1 second; if
    `_is_waiting_for_interrupts` is set to False the loop will exit.

    If an exception occurs while waiting for interrupts, the interrupt
    gpio interfaces will be cleaned up (/sys/class/gpio unexports). In
    this case all interrupts will be reset and you'd need to add the
    callbacks again before using `wait_for_interrupts(..)` again.
    """
    global _is_waiting_for_interrupts
    _is_waiting_for_interrupts = True
    try:
        while _is_waiting_for_interrupts:
            events = _epoll.poll(epoll_timeout)
            for fileno, event in events:
                if event & select.EPOLLPRI:
                    f = _map_fileno_to_file[fileno]
                    # read() is workaround for not getting new values
                    # with read(1)
                    val = f.read().strip()
                    f.seek(0)
                    _handle_interrupt(fileno, val)
    except:
        debug("RPIO: auto-cleaning interfaces after an exception")
        _cleanup_interfaces()
        raise


def stop_waiting_for_interrupts():
    """
    Ends the blocking `wait_for_interrupts()` loop the next time it can,
    which depends on the `epoll_timeout` (per default its 1 second).
    """
    global _is_waiting_for_interrupts
    _is_waiting_for_interrupts = False


def _cleanup_interfaces():
    """
    Remove all /sys/class/gpio/gpioN interfaces that this script created,
    and delete all callback bindings. Should be used after using interrupts.
    """
    debug("Cleaning up interfaces...")
    global _gpio_kernel_interfaces_created
    for gpio_id in _gpio_kernel_interfaces_created:
        # Close the value-file and remove interrupt bindings
        try:
            del_interrupt_callback(gpio_id)
        except:
            pass

        # Remove the kernel GPIO interface
        debug("- unexporting GPIO %s" % gpio_id)
        with open(_SYS_GPIO_ROOT + "unexport", "w") as f:
            f.write("%s" % gpio_id)

    # Reset list of created interfaces
    _gpio_kernel_interfaces_created = []


def cleanup():
    """
    Clean up by resetting all GPIO channels that have been used by this
    program to INPUT with no pullup/pulldown and no event detection. Also
    unexports the interrupt interfaces and callback bindings. You'll need
    to add the interrupt callbacks again before waiting for interrupts again.
    """
    _cleanup_interfaces()
    _cleanup_orig()
